File: antColony.go
package main

import (
    "fmt"
    "math"
    "math/rand"
    "sort"
    "time"
)

const INF = math.MaxInt32
const Q = 1000

func HungarianMethod(matrix [][]int) []int {
    n := len(matrix)
    m := len(matrix[0])

    u := make([]int, n+1)
    v := make([]int, m+1)
    p := make([]int, m+1)
    way := make([]int, m+1)

    for i := 1; i <= n; i++ {
        p[0] = i
        minv := make([]int, m+1)
        used := make([]bool, m+1)
        for j := 0; j <= m; j++ {
            minv[j] = INF
            used[j] = false
        }

        f := 0
        var j0 int
        for {
            used[j0] = true
            i0 := p[j0]
            delta := INF
            j1 := 0
            for j := 1; j <= m; j++ {
                if !used[j] {
                    cur := matrix[i0-1][j-1] - u[i0] - v[j]
                    if cur < minv[j] {
                        minv[j] = cur
                        way[j] = j0
                    }
                    if minv[j] < delta {
                        delta = minv[j]
                        j1 = j
                    }
                }
            }
            if delta == INF {
                break
            }
            for j := 0; j <= m; j++ {
                if used[j] {
                    u[p[j]] += delta
                } else {
                    minv[j] -= delta
                }
                v[j] -= delta
            }
            j0 = j1
            if p[j0] == 0 {
                break
            }
        }
        for {
            j1 := way[j0]
            p[j0] = p[j1]
            j0 = j1
            if j0 == 0 {
                break
            }
        }
    }

    result := make([]int, n)
    for j := 1; j <= m; j++ {
        if p[j] != 0 {
            result[p[j]-1] = j - 1
        }
    }
    return result
}

func SolveTSPWithAntAlgorithm(tspMatrix [][]int, iterations, numAnts int, alpha, beta, evaporationRate float64) (int, []int) {
    rand.Seed(time.Now().UnixNano())
    n := len(tspMatrix)
    if n == 0 {
        return 0, nil
    }

    maxTspValue := 0
    for i := 0; i < n; i++ {
        for j := 0; j < n; j++ {
            if tspMatrix[i][j] != INF && tspMatrix[i][j] > maxTspValue {
                maxTspValue = tspMatrix[i][j]
            }
        }
    }

    cMatrix := make([][]int, n)
    for i := 0; i < n; i++ {
        cMatrix[i] = make([]int, n)
        for j := 0; j < n; j++ {
            if tspMatrix[i][j] == INF {
                cMatrix[i][j] = 10 * maxTspValue
            } else {
                cMatrix[i][j] = tspMatrix[i][j]
            }
        }
    }

    pheromone := make([][]float64, n)
    for i := 0; i < n; i++ {
        pheromone[i] = make([]float64, n)
        for j := 0; j < n; j++ {
            pheromone[i][j] = 1.0
        }
    }

    bestCost := INF
    var bestAssignment []int

    for iter := 0; iter < iterations; iter++ {
        allSolutions := make([]struct {
            cost         int
            assignment   []int
            selectedWorks []int
        }, numAnts)

        for ant := 0; ant < numAnts; ant++ {
            visited := make(map[int]bool)
            currentCity := rand.Intn(n)
            visited[currentCity] = true
            selectedWorks := []int{currentCity}

            for len(selectedWorks) < n {
                probabilities := make([]float64, n)
                total := 0.0

                for nextCity := 0; nextCity < n; nextCity++ {
                    if visited[nextCity] || nextCity == currentCity {
                        probabilities[nextCity] = 0
                    } else {
                        pher := math.Pow(pheromone[currentCity][nextCity], alpha)
                        dist := math.Pow(1.0/float64(cMatrix[currentCity][nextCity]), beta)
                        prob := pher * dist
                        probabilities[nextCity] = prob
                        total += prob
                    }
                }

                if total > 0 {
                    for nextCity := 0; nextCity < n; nextCity++ {
                        probabilities[nextCity] /= total
                    }
                }

                r := rand.Float64()
                sumProb := 0.0
                nextCity := currentCity
                for i := 0; i < n; i++ {
                    sumProb += probabilities[i]
                    if sumProb >= r {
                        nextCity = i
                        break
                    }
                }

                visited[nextCity] = true
                selectedWorks = append(selectedWorks, nextCity)
                currentCity = nextCity
            }

            cSelected := make([][]int, n)
            for i := 0; i < n; i++ {
                cSelected[i] = make([]int, n)
                for j := 0; j < n; j++ {
                    cSelected[i][j] = cMatrix[selectedWorks[i]][selectedWorks[j]]
                }
            }

            rawAssignment := HungarianMethod(cSelected)

            filteredAssignment := make([]int, n)
            copy(filteredAssignment, rawAssignment)

            assigned := make([]bool, n)
            for j := 0; j < n; j++ {
                iAssigned := filteredAssignment[j]
                if iAssigned == j {
                    for k := 0; k < n; k++ {
                        if k != j && !assigned[k] {
                            iAssigned = k
                            assigned[k] = true
                            break
                        }
                    }
                }
                filteredAssignment[j] = iAssigned
                assigned[iAssigned] = true
            }

            totalCost := 0
            for j := 0; j < n; j++ {
                iAssigned := filteredAssignment[j]
                totalCost += cSelected[iAssigned][j]
            }

            allSolutions[ant] = struct {
                cost         int
                assignment   []int
                selectedWorks []int
            }{totalCost, filteredAssignment, selectedWorks}

            if totalCost < bestCost {
                bestCost = totalCost
                bestAssignment = filteredAssignment
            }
        }

        for i := 0; i < n; i++ {
            for j := 0; j < n; j++ {
                pheromone[i][j] *= (1 - evaporationRate)
            }
        }

        for ant := 0; ant < numAnts; ant++ {
            sol := allSolutions[ant]
            cost := sol.cost
            assignment := sol.assignment
            works := sol.selectedWorks

            if cost == 0 {
                continue
            }
            deposit := float64(Q) / float64(cost)

            for j := 0; j < n; j++ {
                iAssigned := assignment[j]
                if iAssigned != j {
                    from := works[iAssigned]
                    to := works[j]
                    pheromone[from][to] += deposit
                }
            }
        }
    }

    return bestCost, bestAssignment
}

File: antColony.py
import numpy as np
import random
from copy import deepcopy

def SolveTSPWithAntAlgorithm(tspMatrix, iterations=100, numAnts=10, alpha=1, beta=2, evaporationRate=0.1, Q=100):
    n = len(tspMatrix)
    INF = float('inf')
    
    pheromone = [[1.0 for _ in range(n)] for _ in range(n)]
    maxTspValue = max(max(val for val in row if val != INF) for row in tspMatrix)
    
    cMatrix = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            tspValue = tspMatrix[i][j]
            if tspValue == INF:
                tspValue = 10 * maxTspValue
            cMatrix[i][j] = tspValue
    
    bestCost = float('inf')
    bestAssignment = None
    
    for iteration in range(iterations):
        all_solutions = []
        
        for ant in range(numAnts):
            visited = set()
            currentCity = random.randint(0, n-1)
            visited.add(currentCity)
            selectedWorks = [currentCity]
            
            while len(selectedWorks) < n:
                probabilities = [0.0] * n
                total = 0.0
                
                for nextCity in range(n):
                    if nextCity in visited or nextCity == currentCity:
                        probabilities[nextCity] = 0.0
                    else:
                        prob = (pheromone[currentCity][nextCity] ** alpha) * \
                              ((1.0 / cMatrix[currentCity][nextCity]) ** beta)
                        probabilities[nextCity] = prob
                        total += prob
                
                if total > 0:
                    probabilities = [p/total for p in probabilities]
                    nextCity = random.choices(range(n), weights=probabilities, k=1)[0]
                else:
                    for city in range(n):
                        if city not in visited and city != currentCity:
                            nextCity = city
                            break
                
                selectedWorks.append(nextCity)
                visited.add(nextCity)
                currentCity = nextCity
            
            C_selected = [[0] * n for _ in range(n)]
            for i in range(n):
                for j in range(n):
                    C_selected[i][j] = cMatrix[selectedWorks[i]][selectedWorks[j]]
            
            rawAssignment = solve_assignment(C_selected)     

            filteredAssignment = rawAssignment.copy()
            assigned = set(filteredAssignment)
            
            for j in range(n):
                i_assigned = filteredAssignment[j]
                if i_assigned == j:
                    for k in range(n):
                        if k != j and k not in assigned:
                            i_assigned = k
                            break
                    filteredAssignment[j] = i_assigned
                    assigned.add(i_assigned)

            totalCost = sum(C_selected[i][j] for j, i in enumerate(filteredAssignment))
            
            all_solutions.append((totalCost, filteredAssignment, selectedWorks))
            
            if totalCost < bestCost:
                bestCost = totalCost
                bestAssignment = filteredAssignment

        for i in range(n):
            for j in range(n):
                pheromone[i][j] *= (1 - evaporationRate)
        
        for solution in all_solutions:
            cost, assignment, works = solution
            if cost == 0: 
                continue
            for j in range(n):
                i_assigned = assignment[j]
                pheromone[works[i_assigned]][works[j]] += Q / cost
    
    return bestCost, bestAssignment


def solve_assignment(cost_matrix):
    n = len(cost_matrix)
    assignment = []
    used_rows = set()
    
    for j in range(n):
        min_cost = float('inf')
        best_i = -1
        for i in range(n):
            if i not in used_rows and cost_matrix[i][j] < min_cost:
                min_cost = cost_matrix[i][j]
                best_i = i
        if best_i == -1:
            for i in range(n):
                if i not in used_rows:
                    best_i = i
                    break
        assignment.append(best_i)
        used_rows.add(best_i)
    return assignment
